src/PerfView/PerfViewData.cs

```diff
@@ -463,11 +463,40 @@ public static string ChangeExtension(string filePath, string newExtension)
             return Path.Combine(dirName, fileName);
         }
 
+        /// <summary>
+        /// for extensions of PerfView that don't have access to the GuiApp.MainWindow
+        /// </summary>
+        /// <param name="doAfter"></param>
+        public void Open(Action doAfter = null)
+        {
+            this.Open(GuiApp.MainWindow, GuiApp.MainWindow.StatusBar, doAfter);
+        }
+
         public override void Open(Window parentWindow, StatusBar worker, Action doAfter = null)
         {
             if (!m_opened)
             {
-                worker.StartWork("Opening " + Name, delegate ()
+                if (worker.IsWorking)
```

Comment:
Please put a comment here that we are trying to support reentrancy (calling Open when processing an Open), and thus can simply do the work inline (since we are already on a worker thread).

Reply:
Hmm, the more I think about it the more concerned I get.   Calls like 'Open' are meant to be run on the GUI thread, but if this is a reentrant case, you are probably NOT on the GUI thread (you are on a worker thread).    This is problematic.  

I would like to understand your scenario better.    Seems like you want access to call OpenImpl (which does all the work that can be done on another non-gui thread), but I really don't know


I would like to know more about your scenario.

Reply:
When we open a .diagsession in PerfView, the Open/OpenImpl of the diagsession implementation extracts the resources embedded in the diagsession container and adds them to the Children collection. The VisualStudioRPS extension is then loaded (it implements OnOpen). I have added code to the extension to get at the embedded Etl resource, cast it as an EtlDataFile and Open it. This is done so that PerfView will process the Etl and hydrate its Children so that the VisualStudioRPS extension can then populate the tree view and related Views. Without this change, the Etl file will only be displayed as a node in the treeview and expanding it (dbl-click) will bypass any of the extension code and related visualizations in PerfView (because by that point we have already returned from OnOpen).

The problem was that I couldn't call EtlDataFile.Open(window, statusbar, action) from w/n the extension as I didn't have a way to access those elements. You had previously suggested adding the Open() overload to address this limitation. I had to leverage the existing worker because calling worker.StartWork resulted in a bunch of exceptions. Testing seemed to validate that it works, but certainly there is a risk that I missed some cases.

Here is the call sequence:
 - .diagsession is opened in PerfView
   - calls DiagSessionPerfViewFile.Open
       - calls DiagSessionPerfViewFile.OpenImpl  <-- enumerates the resources in the diagsession container and adds them to the Childrens collection
       - calls PerfViewFile.ExecuteOnOpenCommand  <-- loads extensions that implemented OnOpen
          - calls VisualStudioRPS.OpenEtlFile <-- extension implementation
             -  gets the target Etl resource from the DiagSessionPerfViewFile container
             - casts the Etl resource as ETLPerfViewData
             - calls ETLPerfViewData.Open  <-- in order to initialize the contents of the ETL data file for display